common/src/main/java/io/pravega/common/util/btree/BTreeIndex.java

```diff
@@ -0,0 +1,825 @@
+        return performUpdate(toDelete, key -> key, BTreePage::delete, timer);
+    }
+
+    /**
+     * Returns an AsyncIterator that will iterate through all the keys within the specified bounds. All iterated keys will
+     * be returned in order (smallest to largest).
+     *
+     * @param firstKey          A ByteArraySegment representing the lower bound of the iteration.
+     * @param firstKeyInclusive If true, firstKey will be included in the iteration (if it exists in the index), otherwise
+     *                          it will not.
+     * @param lastKey           A ByteArraySegment representing the upper bound of the iteration.
+     * @param lastKeyInclusive  If true, lastKey will be included in the iteration (if it exists in the index), otherwise
+     *                          it will not.
+     * @param fetchTimeout      Timeout for each invocation of AsyncIterator.getNext().
+     * @return A new AsyncIterator instance.
+     */
+    public AsyncIterator<List<PageEntry>> iterator(@NonNull ByteArraySegment firstKey, boolean firstKeyInclusive,
+                                                   @NonNull ByteArraySegment lastKey, boolean lastKeyInclusive, Duration fetchTimeout) {
+        ensureInitialized();
+        return new EntryIterator(firstKey, firstKeyInclusive, lastKey, lastKeyInclusive, this::locatePage, this.state.get().length, fetchTimeout);
+    }
+
+    //endregion
+
+    //region Helpers
+
+    /**
+     * Executes the given updates on the index and persists the changes to the external data source.
+     * @param updates An Iterator of the updates to execute. The Iterator must return the updates in sorted order (by key).
+     * @param getKey  A Function that, when applied to an Update, will return the Key which it updates.
+     * @param action  A BiConsumer that, when applied to a BTreePage and a Collection of actions that need to operate on
+     *                that page, will apply the actions to that BTreePage.
+     * @param timer   Timer for the operation.
+     * @param <T>     Type of the action to execute.
+     * @return A CompletableFuture that, when completed normally, will indicate that the updates have been applied
+     * successfully and will contain the current version of the index (any modifications to the index will result in a
+     * larger version value). If the operation failed, the Future will be completed with the appropriate exception.
+     */
+    private <T> CompletableFuture<Long> performUpdate(Iterator<T> updates, Function<T, ByteArraySegment> getKey,
+                                                      BiConsumer<BTreePage, Collection<T>> action, TimeoutTimer timer) {
+        PageCollection pageCollection = new PageCollection(this.state.get().length);
+        AtomicReference<PageWrapper> lastPage = new AtomicReference<>(null);
+        val lastPageUpdates = new ArrayList<T>();
+        return Futures.loop(
+                updates::hasNext,
+                () -> {
```

Comment:
it could be, but it needs access to `lastPage` and `lastPageUpdates` which are shared with the `thenComposeAsync` block below, so I'd have to put both of them in such a method, which wouldn't attain much.